rfcs/0097-unify-packed-and-object.md

```diff
@@ -0,0 +1,677 @@
+
+- **D3a**: AnyValue as container item
+- **D3b**: Turn Everything to Object and use Object* as container Item
+- **D3c**: UnifyItem(Union[Object*, AnyValue*]) as container item. Container contains pod_pool.
+
+D3b would a desirable choice for compilers as it mostly operates on objects and freq small-str overhead may not be an issue. It cost 8 bytes per Object. The main drawback is that when freq small-str and POD is an issue (as being motivated by matx), then we need a different solution.
+
+Both D3a and D3c should be able to handle small-str and POD issue efficiently.
+
+- D3a works well for runtime handling where small-str and POD efficiency can be an issue if the  It will cost 24 bytes(3x of D3a and D3c) per item.
+- D3c’s preserves the overhead of D3a when operating on Object, and cost 32 bytes(1.3x of D3a) when operating on something that is fully small-string and POD.
+
+We would recommend D3c, with a caveat that it is indeed slightly more eng effort(considering the pod_pool). Note that likely we can have a common pod_pool class that generate UnifyItem, and object containers build on top of it.
+
+### D4a and D4b: PackedFunc Convention
+
+- **D4a**: Same API as the currnet one
+- **D4b**: First class support of AnyValue/View
+
+```cpp
+int TVMCPackedFunc(PackedFuncHandle handle,
+                   int num_args, TVMAnyView* values, 
+                   TVMAnyValue* out_value);
+```
+
+We will go with D4b as it enables first class passing of small-str and full benefit of the AnyValue/object system unification.
+
+## Phasing
+
+This section discusses implementation strategy of the proposal. The proposal can be implemented in the following phases:
+
+- M0: Architectural change, AnyValue, AnyView, AnyPad,  alias, type_code segmentation.
+    - Implement D0a, D1a, D1section-b, D2b and D4b.
+    - The code change would mostly be conventional changes.
+    - Note that this implies (by intention) change the ABI of packed func. We will update compiler/runtime to do so.
+    - We will introduce adapter to support the TVMArgs during the transition, but favors moving to new state to reduce overall complexity.
+- M1: Introduce new string support (with small string)
+- M2: Introduce universal container
+
+We believe the overall milestones are positive given the net gain obtained to enable pre-processing and stronger interpolation with ML ecosystem and the community as well. It also opens doors to bring in features in projects like matx so we can enable efficient NLP preprocessing together with ML workload in the same pipeline.
+
+Additionally, Unifying FFI and Object would bring further unification and reduction in our overall code complexity while leveling up the extensibility, so it serves as a strong improvement to the overall quality of the project. 
+
+## Drawbacks
```

Comment:
Can you comment on _how_ specifically the proposed change impacts the TVM runtime, e.g. is this a breaking change in some aspect?

Reply:
Please see M0 section in the Phasing part. WIll also add a reference here:

- Note that this implies (by intention) change the ABI of packed func. We will update compiler/runtime to do so.
- We will introduce adapter to support the TVMArgs during the transition, but favors moving to new state to reduce overall complexity.

More details can be found in D4